src/common/task_runner.h

```diff
@@ -42,13 +42,7 @@ class TaskRunner {
 
   TaskRunner(const TaskRunner&) = delete;
   TaskRunner& operator=(const TaskRunner&) = delete;
-
-  ~TaskRunner() {
-    if (state_ != Stopped) {
-      Cancel();
-      auto _ = Join();
-    }
-  }
+  ~TaskRunner() = default;
```

Comment:
Seems we need some guarantee for this object? i.e. the behavior will be undefined while the taskrunner is destructed and the attached threads are still running.

Can refer to std::jthread (https://en.cppreference.com/w/cpp/thread/jthread/~jthread) and std::thread (https://en.cppreference.com/w/cpp/thread/thread/~thread).

Reply:
The data race may be caused by the task runner's `Join` function will clear the concurrent queue, so it could be fixed after moving to the last one to join. It's good for me to keep the Cancel and Join in dctor.